users/engler/fairseq/training.py

```diff
@@ -0,0 +1,162 @@
+        :param int|float mem_rqmt: Memory requirements (per GPU)
+        :param int cpu_rqmt: Required number of CPUs (per GPU)
+        :param int gpu_rqmt: Number of required GPUs
+        :param Path|str python_exe: File path to the executable for running python
+        :param Path|str fairseq_hydra_exe: File path to the Fairseq executable: fairseq-hydra-train (usually in the same folder as python exe '.../bin/')
+        :param Path|str fairseq_git: File path to the fairseq git for alternative call of fairseq-hydra-train (no need to install fairseq here)
+
+        """
+        self.fairseq_hydra_config = fairseq_hydra_config
+        self.command_line_args = command_line_args or []
+
+        stored_epochs = list(range(save_interval, max_epoch, save_interval)) + [
+            max_epoch
+        ]
+        if keep_epochs is None:
+            self.keep_epochs = set(stored_epochs)
+        else:
+            self.keep_epochs = set(keep_epochs)
+        self.max_epoch = max_epoch
+        self.save_interval = save_interval
+        self.out_fairseq_hydra_yaml = self.output_path("fairseq_hydra_config.yaml")
+        self.out_checkpoint_dir = self.output_path("checkpoints", directory=True)
+        self.out_models = {
+            k: PytorchHydraModel(
+                self.out_fairseq_hydra_yaml,
+                self.output_path("checkpoints/checkpoint{}.pt".format(k)),
+                k,
+            )
+            for k in stored_epochs
+            if k in self.keep_epochs
+        }
+        self.gpu_rqmt = gpu_rqmt
+        self.fairseq_python_exe = fairseq_python_exe
+        self.fairseq_hydra_exe = (
+            fairseq_hydra_exe if fairseq_hydra_exe is not None else gs.FAIRSEQ_HYDRA_EXE
+        )
+        self.fairseq_git = fairseq_git
+
+        self.rqmt = {
+            "gpu": gpu_rqmt,
+            "cpu": cpu_rqmt,
+            "mem": mem_rqmt,
+            "time": time_rqmt,
+        }
+
+        if self.gpu_rqmt > 1:
+            self.rqmt["cpu"] *= self.gpu_rqmt
+            self.rqmt["mem"] *= self.gpu_rqmt
+
+    def tasks(self):
+        yield Task("create_files", mini_task=True)
+        yield Task("run", resume="run", rqmt=self.rqmt)
+
+    def create_files(self):
+        self.fairseq_hydra_config.write(self.out_fairseq_hydra_yaml.get_path())
+        util.create_executable("fairseq.sh", self._get_run_cmd())
+
+    def run(self):
+        my_env = os.environ
+        if self.fairseq_git is not None:
+            my_env["PYTHONPATH"] = self.fairseq_git
+        sp.check_call(self._get_run_cmd(), env=my_env)
+
+    def _get_run_cmd(self):
+        run_cmd = [
+            "--config-dir",
+            os.path.dirname(self.out_fairseq_hydra_yaml.get_path()),
+            "--config-name",
+            os.path.basename(self.out_fairseq_hydra_yaml.get_path()),
+        ]
+        run_cmd += self.command_line_args
+        run_cmd += ["checkpoint.save_dir=" + str(self.out_checkpoint_dir)]
```

Comment:
please use `.get_path()` instead of string conversion for all variables which are definitely `Path` or `Variable` objects you want to resolve. This is just a convention we want to introduce to make clear that something is a Path object, and to prohibit any caching mistakes. The implicit conversion `__str__` representation was indended for different purposes orignally, and resolves to `get_cached_path()` and not `get_path()`, which can potentially lead to errors.

Reply:
Of course in the cases you definitely want to have the caching, use `get_cached_path`.